store/postgres/src/deployment_store.rs

```diff
@@ -357,52 +357,111 @@ impl DeploymentStore {
     ) -> Result<i32, StoreError> {
+        use EntityModification::*;
         let mut count = 0;
 
-        for modification in mods {
-            use EntityModification::*;
-
-            let n = match modification {
+        // Group `Insert`s and `Overwrite`s by key, and accumulate `Remove`s.
+        let mut inserts = HashMap::new();
+        let mut overwrites = HashMap::new();
+        let mut removals = HashMap::new();
+        for modification in mods.into_iter() {
+            match modification {
+                Insert { key, data } => {
+                    inserts
+                        .entry(key.entity_type.clone())
+                        .or_insert_with(Vec::new)
+                        .push((key, data));
+                }
                 Overwrite { key, data } => {
-                    let section = stopwatch.start_section("check_interface_entity_uniqueness");
-                    self.check_interface_entity_uniqueness(conn, layout, &key)?;
-                    section.end();
-
-                    let _section = stopwatch.start_section("apply_entity_modifications_update");
-                    layout
-                        .update(conn, &key, data, block_number(ptr))
-                        .map(|_| 0)
+                    overwrites
+                        .entry(key.entity_type.clone())
+                        .or_insert_with(Vec::new)
+                        .push((key, data));
                 }
-                Insert { key, data } => {
-                    let section = stopwatch.start_section("check_interface_entity_uniqueness");
-                    self.check_interface_entity_uniqueness(conn, layout, &key)?;
-                    section.end();
-
-                    let _section = stopwatch.start_section("apply_entity_modifications_insert");
-                    layout
-                        .insert(conn, &key, data, block_number(ptr))
-                        .map(|_| 1)
+                Remove { key } => {
+                    removals
+                        .entry(key.entity_type.clone())
+                        .or_insert_with(Vec::new)
+                        .push(key);
                 }
-                Remove { key } => layout
-                    .delete(conn, &key, block_number(ptr))
-                    // This conversion is ok since n will only be 0 or 1
-                    .map(|n| -(n as i32))
-                    .map_err(|e| {
-                        anyhow!(
-                            "Failed to remove entity ({}, {}, {}): {}",
-                            key.subgraph_id,
-                            key.entity_type,
-                            key.entity_id,
-                            e
-                        )
-                        .into()
-                    }),
-            }?;
-            count += n;
+            }
+        }
+        // apply modification groups
+        for (entity_type, entities) in inserts.iter_mut() {
```

Comment:
I haven't tried it, but would some of the complications around `Vec<&EntityKey>` vs `&Vec<EntityKey>` in `Layout.update` and `Layout.delete` etc. go away if you used `inserts.drain()` here instead of `iter_mut()`? If it does help, maybe we can get avoid building an auxiliary `Vec` in those methods.

Reply:
Let's change the ClampQuery signature so it accepts the same type of vector as the InsertQuery.

Reply:
On second thought, this might not be possible since the `Remove` variant does not have an `Entity` value, but only an `EntityKey`

Reply:
I couldn't figure out how to avoid creating the vector you've mentioned, but I was able to avoid creating a new vector inside `ClampRangeQuery` implementation of `walk_ast` on commit 59f1606c28d72cedc4f6e3b6fa8fecb151d7ba40 by generalizing the `ForeignKeyClauses::is_in` function to accept a vector of either `&str` or `String` So we can use that very same vector.

Reply:
Nice! Yes, it looks like we have to allocate the extra vector no matter what.